### PR TITLE
[SeparateConstOffsetFromGEP] Support multiple indices in reorderGEP

### PR DESCRIPTION
**Pull request by nikic**
The code was essentially already ready to handle multiple indices -- we only need to adjust the non-negative index check to check all indices, instead of only the first one.

### CODE REVIEW

**Comment by nikic** [comment]
Putting this up in the hope that it helps with https://github.com/llvm/llvm-project/pull/68882#issuecomment-2108684748.

I've fixed a bug in the inbounds propagation I noticed while writing this in https://github.com/llvm/llvm-project/commit/b4d1a606c7492d827aff6ff0c1c109adff1253b9.

File: llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp [hunk 1]
```diff
@@ -997,14 +992,12 @@ bool SeparateConstOffsetFromGEP::reorderGEP(GetElementPtrInst *GEP,
...
   if (IsChainInBounds) {
-    auto GEPIdx = GEP->indices().begin();
-    auto KnownGEPIdx = computeKnownBits(GEPIdx->get(), *DL);
-    IsChainInBounds &= KnownGEPIdx.isNonNegative();
-    if (IsChainInBounds) {
-      auto PtrGEPIdx = PtrGEP->indices().begin();
-      auto KnownPtrGEPIdx = computeKnownBits(PtrGEPIdx->get(), *DL);
-      IsChainInBounds &= KnownPtrGEPIdx.isNonNegative();
-    }
+    auto IsKnownNonNegative = [&](Value *V) {
```
Comment (arsenm) [inline llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:995]: DL is the only capture, so capture by value? Is it worth providing the context instruction to the SimplifyQuery?

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp [hunk 2]
```diff
@@ -997,14 +992,12 @@ bool SeparateConstOffsetFromGEP::reorderGEP(GetElementPtrInst *GEP,
...
+      return isKnownNonNegative(V, *DL);
+    };
+    IsChainInBounds &= all_of(GEP->indices(), IsKnownNonNegative);
+    if (IsChainInBounds)
+      IsChainInBounds &= all_of(PtrGEP->indices(), IsKnownNonNegative);
```
Comment (jayfoad) [inline llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:1000]: Does this work? I'm never sure how flexible `concat` is.
```suggestion
    IsChainInBounds &= all_of(concat(GEP->indices(), PtrGEP->indices()), IsKnownNonNegative);
```

**Comment by sgundapa** [comment]
> Putting this up in the hope that it helps with [#68882 (comment)](https://github.com/llvm/llvm-project/pull/68882#issuecomment-2108684748).

I was about to push a similar change.
It did solve the regression I have seen.  Thanks for the patch.

**Review by sgundapa — APPROVED** [review]

**Pull request merged**